### DISCUSSION TITLE
Expiring Presigned URLs

### DISCUSSION THREAD

**Discussion by congenica-peter-fox** [GMOD/jbrowse-components#4740] [Q&A]
Hi, 

We're looking to embedded jbrowse as a react component in an existing application.

We have assembly and sequence data available using pre-signed URLs from an AWS S3 bucket. The pre-signed URLs will expire about 15 minutes after they have been generated. 

I'm looking for a way for refresh the URLs periodically or refresh on demand. Looking at documentation and tutorials I've not found a way of doing this. I was wondering if there was a mechanism where a function could be called when requesting data. 

Thanks
Pete

**Comment by cmdcolin** [comment]
hi there, I have been meaning to make an example of this for awhile (older issue https://github.com/GMOD/jbrowse-components/issues/2744)
 
one option for this is to use our so-called 'internet accounts' system

this allows you to implement a custom 'fetch' routine that can allow a token to be refreshed (our "oauth" internet account system does this with a "refresh token")

there are not a lot of docs about creating your own 'internet account' but it is an element of our plugin system, and creating plugins when using embedded can be somewhat easy (see "plugins defined in your code" from https://jbrowse.org/storybook/lgv/main/?path=/docs/using-plugins--docs, don't have to use the jbrowse-plugin-template or anything)

**Reply by congenica-peter-fox** [reply]
Thanks for the response. Our API's current provide pre-signed URLs whereas the Internet account allow the setting of header values. We could add new APIs to provide a URL to the desired file and use the AWS STS service to provide appropriate short term tokens to access the a file.

The implementation would be slightly more complex as each file would have its own token, so we would need a cache of tokens based on the file URL.

So I can see a path to being able to access files in S3 using the InternetAccount object, but we'd wind up with an implementation which is a bit more tied to AWS than we'd ideally want.

**Reply by cmdcolin** [reply]
if you can come up with a system that would work on a nonjbrowse page with just javascript and fetch, then it is likely that it could be adapted to work with jbrowse. I just asked google gemini to design a system that would do something like this but it could be more complicated than it needs to be or misses the mark in other ways https://g.co/gemini/share/a83f05a910e0
